### PR TITLE
[mlir][EmitC] Do not inline expressions used by ops with the CExpression trait

### LINKED ISSUE #93470: [mlir][emitc] Parenthesization issue in TranslateToCpp
This MLIR: 

```mlir
module {
  func.func @shift() {
    %0 = "emitc.constant"() <{value = 160 : ui32}> : () -> ui32
    %1 = "emitc.constant"() <{value = 7 : ui32}> : () -> ui32
    %2 = emitc.cmp lt, %0, %1 : (ui32, ui32) -> i1
    %3 = emitc.expression : ui32 {
      %4 = emitc.bitwise_right_shift %0, %1 : (ui32, ui32) -> ui32
      %5 = emitc.conditional %2, %4, %0 : ui32
      emitc.yield %5 : ui32
    }
    %6 = emitc.cast %3 : ui32 to i32
    return
  }
}
```

gets translated into this C++ code using `mlir-translate --mlir-to-cpp`:
```c++
void shift() {
  uint32_t v1 = 160;
  uint32_t v2 = 7;
  bool v3 = v1 < v2;
  int32_t v4 = (int32_t) v3 ? v1 >> v2 : v1;
  return;
}
```

where we would expect:
```c++
  int32_t v4 = (int32_t)(v3 ? v1 >> v2 : v1);
```

or
```c++
  uint32_t v4 = v3 ? v1 >> v2 : v1;
  int32_t v5 = (int32_t) v4;
```

The `CppEmitter` needs to emit parentheses around casted values. Here, it doesn't and the cast happens on the predicate, which is incorrect. Fortunately, this looks pretty easy to fix :)

### PR DESCRIPTION
**Pull request by simon-camp**
Currently an expression is inlined without emitting enclosing parentheses regardless of the context of the user. This could led to wrong evaluation order depending on the precedence of both expressions. If the inlining is intended, the user operation should be merged into the expression op.

Fixes #93470.

### CODE REVIEW

**Review by mgehre-amd — APPROVED** [review]
Thank you!

**Comment by aniragil** [comment]
I wonder if we could instead add another restriction to `shouldBeInlined`: Don't inline if the user has the `CExpression` trait. 
The rationale is that the using `CExpression` could have been fused into its operand `ExpressionOp`. If it wasn't, then the pass(es) that formed the expressions did not want it to be fused. Inlining the `ExpressionOp` operand into it goes against that intention.
This would also make sure that we handle expression emission in one place, i.e. `ExpressionOp`s, where `cast` should already be emitted correctly with parenthesis as needed.

**Review by cferry-AMD — APPROVED** [review]
Looks good, thanks :)

**Comment by simon-camp** [comment]
> I wonder if we could instead add another restriction to `shouldBeInlined`: Don't inline if the user has the `CExpression` trait. The rationale is that the using `CExpression` could have been fused into its operand `ExpressionOp`. If it wasn't, then the pass(es) that formed the expressions did not want it to be fused. Inlining the `ExpressionOp` operand into it goes against that intention. This would also make sure that we handle expression emission in one place, i.e. `ExpressionOp`s, where `cast` should already be emitted correctly with parenthesis as needed.

I implemented this in the last commit. I don't know which of these solutions is preferable.

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp [hunk 1]
```diff
@@ -1338,8 +1343,9 @@ LogicalResult CppEmitter::emitOperand(Value value) {
   }
 
   auto expressionOp = dyn_cast_if_present<ExpressionOp>(value.getDefiningOp());
-  if (expressionOp && shouldBeInlined(expressionOp))
+  if (expressionOp && shouldBeInlined(expressionOp)) {
     return emitExpression(expressionOp);
+  }
```
Comment (cferry-AMD) [inline mlir/lib/Target/Cpp/TranslateToCpp.cpp:1348]:
```suggestion
  if (expressionOp && shouldBeInlined(expressionOp))
    return emitExpression(expressionOp);
```

**Review by cferry-AMD — APPROVED** [review]

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp [hunk 2]
```diff
@@ -1338,8 +1343,9 @@ LogicalResult CppEmitter::emitOperand(Value value) {
...
```
Comment (aniragil) [inline mlir/lib/Target/Cpp/TranslateToCpp.cpp:1349]: nit: please restore the [single statement format](https://www.llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements)

**Review by aniragil — CHANGES_REQUESTED** [review]
LGTM (with a nit)

File: mlir/test/Target/Cpp/expressions.mlir [hunk 3]
```diff
@@ -100,6 +100,86 @@ func.func @parentheses_for_same_precedence(%arg0: i32, %arg1: i32, %arg2: i32) -
   return %e : i32
 }
 
+// CPP-DEFAULT:      int32_t user_with_expression_trait(int32_t [[VAL_1:v[0-9]+]], int32_t [[VAL_2:v[0-9]+]], int32_t [[VAL_3:v[0-9]+]]) {
+// CPP-DEFAULT-NEXT:   int32_t v4 = 0;
```
Comment (TinaAMD) [inline mlir/test/Target/Cpp/expressions.mlir:104]: Nit: Can you capture all the local variables?
```suggestion
// CPP-DEFAULT-NEXT:   int32_t [[VAL_4:v[0-9]+]] = 0;
```

**Review by TinaAMD — COMMENTED** [review]
Thanks for taking care of this!

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp [hunk 4]
```diff
@@ -303,7 +303,12 @@ static bool shouldBeInlined(ExpressionOp expressionOp) {
 
   // Do not inline expressions used by other expressions, as any desired
   // expression folding was taken care of by transformations.
-  return !user->getParentOfType<ExpressionOp>();
+  if (user->getParentOfType<ExpressionOp>())
+    return false;
+
+  // Do not inline expressions used by ops with the CExpression trait. If this
+  // was intended, the user could have been merged into the expression op.
+  return !user->hasTrait<OpTrait::emitc::CExpression>();
```
Comment (aniragil) [inline mlir/lib/Target/Cpp/TranslateToCpp.cpp:311]: Actually, isn't the first check now redundant?
If the user is part of an ExpressionOp then it must have the CExpression trait, right?

**Review by aniragil — CHANGES_REQUESTED** [review]

**Review by aniragil — APPROVED** [review]
Thanks @simon-camp !

**Review by TinaAMD — APPROVED** [review]
LGTM

**Pull request merged**